Sources/SwiftDocC/Infrastructure/ContentCache.swift

```diff
@@ -0,0 +1,93 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2024 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See https://swift.org/LICENSE.txt for license information
+ See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+extension DocumentationContext {
+    /// A cache for symbol and page content.
+    struct ContentCache<Value> {
+        private var storage = [ResolvedTopicReference: Value]()
+        private var symbolIndex = [String: ResolvedTopicReference]()
+        
+        /// Accesses the value for a given reference.
+        /// - Parameter reference: The reference to find in the cache.
+        subscript(reference: ResolvedTopicReference) -> Value? {
+            // Avoid copying the values if possible
+            _read { yield storage[reference] }
+            _modify { yield &storage[reference] }
+        }
+        
+        /// Adds a value to the cache for a given reference _and_ symbol ID.
+        /// - Parameters:
+        ///   - value: The value to add to the cache.
+        ///   - reference: The reference associated with that value.
+        ///   - symbolID: The symbol ID associated with that value.
+        mutating func add(value: Value, reference: ResolvedTopicReference, symbolID: String) {
+            symbolIndex[symbolID] = reference
```

Comment:
Even overlapping writes to a single dictionary isn't safe. For example, the code below will crash even though every write operation has a unique key:
```swift
var storage = [Int: Int]()
DispatchQueue.concurrentPerform(iterations: 2) { iteration in
    for number in 0..<10_000 {
        storage[(iteration * 10_000) + number] = number
    }
}
```

This means that the lack of an atomic operation here isn't a real concern because any multithreaded mutation  would crash regardless. If the number of writes is very small it's off course possible to not crash consistently but the thread sanitizer can help find such unsafe overlapping writes regardless.

This was true even before wrapping these dictionaries, so the `DocumentationContext` has a pattern during content registration where expensive operations are performed in parallel—gathering results in separate variables—and then the main stored properties are updated synchronously. After all the content is registered, the context is no longer mutated—although its variables are still mutate<em>able</em>—so the rendering process can safely _read_ from the context from multiple threads. Ideally, the context would be made fully immutable during rendering so that it's not possible to accidentally mutate it during rendering. We're tracking that under-the-hood-change as <rdar://65130130>.

----

I can add a few notes in the documentation comments to say that mutating access isn't thread safe.

Reply:
I added a note about the lack of thread-safety in https://github.com/apple/swift-docc/pull/806/commits/7322841116333dfb895746d347a31fd213eb147e